### PR TITLE
Added Particle Emitter Pause option, and tests

### PR DESCRIPTION
**Pull request by PooperPig**
Added methods to Pause and UnPause emissions.
When paused, existing particles continue but new particles are not created.
UnPause continues to emit new particles.
Also adds IsPaused boolean.

Included Test/Demo of the effect in the tests.

I accidentally included changes to CCParticleSystemQuad (I made changes but then undid them - and did not intend to include them in the pull.

### CODE REVIEW

File: cocos/2d/CCParticleSystem.h [hunk 1]
```diff
@@ -801,6 +801,17 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
     
     //! Initializes a system with a fixed number of particles
     virtual bool initWithTotalParticles(int numberOfParticles);
+    
+    /** Are the emissions paused
+     @return True if the emissions are paused, else false
+     */
+    virtual bool isPaused();
```
Comment (minggo) [inline cocos/2d/CCParticleSystem.h]: virtual bool isPaused() -> virtual bool isPaused() const?

File: cocos/2d/CCParticleSystem.h [hunk 2]
```diff
@@ -801,6 +801,17 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
...
+    
+    /* Pause the emissions*/
+    virtual void pauseEmissions();
+    
+    /* UnPause the emissions*/
+    virtual void unPauseEmissions();
```
Comment (minggo) [inline cocos/2d/CCParticleSystem.h]: How about rename to `resumeEmissions()` ?

File: cocos/2d/CCParticleSystem.h [hunk 3]
```diff
@@ -956,6 +967,9 @@ class CC_DLL ParticleSystem : public Node, public TextureProtocol, public Playab
      @since v0.8
      */
     PositionType _positionType;
+    
+    /** is the emitter paused */
+    bool paused;
```
Comment (minggo) [inline cocos/2d/CCParticleSystem.h]: paused -> _ paused to according cocos2d-x coding style

File: cocos/2d/CCParticleSystemQuad.cpp [hunk 4]
```diff
@@ -707,4 +707,10 @@ std::string ParticleSystemQuad::getDescription() const
     return StringUtils::format("<ParticleSystemQuad | Tag = %d, Total Particles = %d>", _tag, _totalParticles);
 }
 
+
+
+
+
+
+
```
Comment (minggo) [inline cocos/2d/CCParticleSystemQuad.cpp]: It is better not add unneeded spaces.

File: cocos/2d/CCParticleSystemQuad.h [hunk 5]
```diff
@@ -144,6 +144,8 @@ class CC_DLL ParticleSystemQuad : public ParticleSystem
 
     virtual std::string getDescription() const override;
     
+
+
```
Comment (minggo) [inline cocos/2d/CCParticleSystemQuad.h]: Please don't add unneeded spaces.

**Comment by minggo** [comment]
Thanks.

**Comment by PooperPig** [comment]
No worries - thank you!

Max Young
mailmaxxx@gmail.com

File: cocos/2d/CCParticleSystem.cpp [hunk 6]
```diff
@@ -588,6 +588,8 @@ ParticleSystem::~ParticleSystem()
 
 void ParticleSystem::addParticles(int count)
 {
+    if (_paused)
```
Comment (dumganhar) [inline cocos/2d/CCParticleSystem.cpp:591]: Why `_pause` wasn't initialized to false? It will be a random value (true / false).

Comment (dumganhar) [inline cocos/2d/CCParticleSystem.cpp:591]: fix initialization problem at https://github.com/cocos2d/cocos2d-x/pull/15920

**Pull request merged**